pkg/hubble/proxy/server.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+
+	"github.com/cilium/hubble/api/v1/observer"
+	v1 "github.com/cilium/hubble/pkg/api/v1"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/health"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/status"
+)
+
+// Server is a proxy that connects to a running instance of hubble gRPC server
+// via unix domain socket.
+type Server struct {
+	client observer.ObserverClient
+	server *grpc.Server
+	opts   Options
+}
+
+// NewServer creates a new proxy Server.
+func NewServer(options ...Option) (*Server, error) {
+	opts := DefaultOptions
+	for _, opt := range options {
+		if err := opt(&opts); err != nil {
+			return nil, fmt.Errorf("failed to apply option: %v", err)
+		}
+	}
+	return &Server{
+		opts: opts,
+	}, nil
+}
+
+// ensure that GRPCServer implements the observer.ObserverServer interface.
+var _ observer.ObserverServer = (*Server)(nil)
+
+// Serve starts the hubble proxy server.
+func (s *Server) Serve() error {
+	if err := s.connectClient(); err != nil {
+		return fmt.Errorf("failed to connect to hubble instance: %v", err)
+	}
+
+	s.server = grpc.NewServer()
+	healthSrv := health.NewServer()
+	healthSrv.SetServingStatus(v1.ObserverServiceName, healthpb.HealthCheckResponse_SERVING)
+	healthpb.RegisterHealthServer(s.server, healthSrv)
+	observer.RegisterObserverServer(s.server, s)
+	socket, err := net.Listen("tcp", s.opts.ListenAddress)
+	if err != nil {
+		return fmt.Errorf("failed to listen on tcp socket %s: %v", s.opts.ListenAddress, err)
+	}
+
+	s.server.Serve(socket)
+	return nil
+}
+
+func (s *Server) connectClient() error {
+	ctx, cancel := context.WithTimeout(context.Background(), s.opts.DialTimeout)
+	defer cancel()
```

Comment:
this defer will close the connection right after the function exits

Reply:
the doc says that the context only controls the setup and not the connection: https://godoc.org/google.golang.org/grpc#DialContext @Rolinh i guess you tried it and it worked?

Reply:
> In the blocking case, ctx can be used to cancel or expire the pending connection. Once this function returns, the cancellation and expiration of ctx will be noop. Users should call ClientConn.Close to terminate all the pending operations after this function returns. 


it should be ok.

Reply:
Yeah since it's a blocking connection looks ok.

Reply:
@michi-covalent Yup, of course I tried it and yes it works. I extracted this code out of `Serve()` which is why it's the way it is but it would make sense to pass the context around and `defer cancel()` as part of `Serve()` anyway.

Reply:
Actually no, it's correct the way it is as this context is just for the dial timeout.